Add xxh64 and fingerprint call-caching strategies to address shortcomings of current strategies. [BA-6335]

rhpvorderman opened the pull request:
## Call-caching problems with path+modtime
I have been doing some call-caching benchmarking on the [BioWDL RNA-seq](https://github.com/biowdl/RNA-seq) pipeline and it turns out any `path` or `path+modtime` strategies do not work with containers. As is reported in these issues: #5405, #5370, #5346 . @cmarkello, @illusional, I am sorry that I insisted that `path+modtime` did work. I was using less complex workflows that did not have this problem at the time.

## Call-caching problems with file strategy
The `file` strategy does work as it uses md5sums in order to calculate the file hash. An unfortunate side effect of this is that md5 uses massive system resources. On HPC systems that are the target for the sfs-backend, this is a big problem. Cromwell will be run from a submit node on the system and greedily grab all processing power on the submit node to calculate all the md5sums. 

## Md5sums
Md5sums are reliable hashes for file integrity, but this was not their intended purpose. Md5sum was intended as a cryptographic hash. A cryptographic hash has the following properties (wikipedia):
1. it is deterministic, meaning that the same message always results in the same hash
2. it is quick to compute the hash value for any given message
3. it is infeasible to generate a message that yields a given hash value
4. it is infeasible to find two different messages with the same hash value
5. a small change to a message should change the hash value so extensively that the new hash value appears uncorrelated with the old hash value (avalanche effect)

I contest point 2, in that many cryptographic explicitly strife for being slow to calculate in order to negate brute force attempts.
Anyway: for call caching we only need points 1. and 4. All the rest is unnecessary ballast. 

## xxHash
Luckily there is a hashing algorithm that is designed explicitly for content hashing only. It was made to generate reliably different hashes for file content as fast as possible. It's called [xxHash](https://www.xxhash.com).

There are Java implementations available and I did [some extensive benchmarking](https://github.com/rhpvorderman/hashtest/) to find out which one was best. The xxh64 (xxhash for 64 bit machines) algorithm was 15 times faster than the java implementation of md5 we currently use in Cromwell.

This PR implements the xxhash algorithm for call-caching in Cromwell:

+ The default strategy will still be using md5 for backwards compatibility.
+ A new `xxh64` strategy is implemented using the 64-bit xxhash algorithm. (I didn't make the xxh32 algorithm available. Is there any Cromwell server still running on 32-bit?) This can be set in the call caching configuration.
+ A new `fingerprint` strategy suggested by @illusional, which takes the modtime, size and a xxh64 hash of the first 10 mb of the file to create a virtually unique fingerprint.
+ The `file` strategy get's a new alias `md5` which is more clear. Although `file` will still work in the config for backwards compatibility. 

I feel we should move to xxh64 as default after it has proven itself in a few releases. The speed-up is an order of magnitude.

rhpvorderman commented:
I have added the "DO NOT MERGE" label because I want to test this change on our research cluster first. I will do this next week. I have a 100 sample run that needs to be reanalyzed using a stable version of the RNA-seq pipeline. This spawns some 2000ish jobs. So I can really put the new strategy trough its paces and check if it works correctly. I will report on this next week and remove the "DO NOT MERGE" label if everything turns out to be okay.

rhpvorderman commented on CHANGELOG.md:
```diff
@@ -1,5 +1,22 @@
+
+### Bugfixes
+
++ Fixed a bug in Cromwell 49 where the `use_relative_output_paths` option would not preserve intermediate folders.
```
This was another PR of mine for which I forget to update the changelog :man_facepalming: , sorry!

rhpvorderman replied:
Moved it to another PR.

cjllanwarne commented on CHANGELOG.md:
```diff
@@ -1,5 +1,22 @@
+
+### New functionality
+
+#### xxHash for faster call-caching using file hashes
```
Rather than putting all the detail in the changelog, could you document this in user facing docs (eg readthedocs) and then put a brief placeholder linking to that from here?

Otherwise, in a couple of versions time, it'll be really hard to direct anyone to the documentation for this feature.

illusional commented:
I thought I was going crazy! Thanks @rhpvorderman, I think this is a great PR, and really looking forward to being able to test this on our HPC. I'm still a little concerned with Cromwell requesting the whole file (up to 200GB) and it's impact on our network. But this is definitely a step in the write direction.

I was thinking of maybe adding something similar, but would check the file size, and then perform `xxh64` on the first 1MB or some region of the file (eg: `size+xxh64-partial`).

I'm happy to produce some documentation (whether a part of this PR, or linked to it), about call-caching on Cromwell and to include this new material once I can test it okay. Thanks for pinging me @rhpvorderman, excited to try this out!

rhpvorderman commented:
@illusional . A partial hash is a great idea. I am now downloading a 82 GB bam file in order to check the speed of the algorithm, but unfortunately my network connection is a 100mbits on this PC. It will take 2 hours. Even with a 1000mbit perfect connection it would take at least 12 minutes. So computation time is indeed not the limit here.

We need to thinks this through though. Some files are more similar at the beginning (VCF headers come to mind) than at the end. So only hashing the beginning carries with it some major concerns. Using the size is indeed a good thing, and I think we should also include the modification time.
In that case `size+modtime+xx64hsum of first 10mb` should create a unique enough identifier for each file while negating any network performance issues. I think this strategy should be called `hpc`.

rhpvorderman commented:
@illusional. I did some benchmarking on xxh64sum vs md5sum on a 35GB file. Results:
* Just reading the file with `cat <file> /dev/null` took 18 seconds. Virtually no CPU time
* xxh64sum took 24 seconds of which 3.6 seconds cpu time
* md5sum took 53 seconds, of which 48 seconds cpu time.

Md5sum was cpu limited. So CPU was 100% all the time. xxh64sum was limited by the transfer speed of the disk (nvme ssd), so cpu usage never exceeded 20%. This means the bottleneck becomes I/O based, and for 200 GB files on NFS this can indeed be a big problem.

I have added a `hpc` strategy` that takes the last modified time, size, and the xxh64sum of the first 10 megabytes of the file to alleviate this problem.

illusional commented:
You're awesome @rhpvorderman, can't wait to try this out. I'm not the best crash hot at Scala, but is it enough for me to check out your branch, `sbt` to get the jar and test this?

rhpvorderman commented:
@illusional. I renamed the strategy `fingerprint` because I think it can be used in a general case. Also because it is called "fingerprint" it does carry with it the sense that it only tests a small part of the file, and is therefore less reliable than a strategy that hashes the entire file. (Even though it should be reliable enough).

To build a new jar, check out the [documentation](https://cromwell.readthedocs.io/en/stable/developers/Building/). It is as easy indeed as checking out the branch and running `sbt assembly`. It might take a while though. If you run out of memory I believe sbt has a `-mem` flag to set the memory.

@cjllanwarne I fully agree with your comments on the documentation part, so I trimmed the changelog and moved the information to the documentation. I hope the documentation is adequate and well-explained enough.

rhpvorderman commented:
@illusional I have also provided jars here: https://github.com/rhpvorderman/cromwell/releases/tag/50-dev-lumctest

rhpvorderman commented:
I have run both strategies with a workflow that generates around about 2000 jobs (100 samples) using GATK best practices for RNA variant calling. 

### Method

The Cromwell instance ran with a SLURM cluster backend. All jobs were run using singularity containers. The cromwell process was limited to 3 akka threads and 1 GC thread (by default it grabs al threads on the login node, and this is not fair to other users). The HSQLDB memory database with persistance file was used. Said SLURM cluster has its storage connected via NFS.

Two configurations of cromwell were used. One with the xxh64 strategy, and one with the fingerprint strategy. Each cromwell instance was executed in its own directory, with its own database and own cromwell-executions folder. The [BioWDL RNA-seq](https://github.com/biowdl/rna-seq) workflow was run. After running, the workflow was run again to see if the call-caching worked correctly.

### Results
Both `xxh64` and `fingerprint` strategies were able to rerun the workflow with a 100% Cache hit. The fingerprint strategy however was much quicker:
`time` results for fingerprint
```
real    23m26.269s
user    15m31.229s
sys     2m43.406s
```
`time` results for xxh64
```
real    69m12.478s
user    56m7.371s
sys     52m6.262s
```

### Conclusion
Using xxh64 as a strategy requires some calculation but one hour for 100 samples on 2000 jobs is quite acceptable. What is obvious is that the system IO (`sys` time) takes a lot of time as well. This cluster has very fast optimized ISILON storage, but on clusters without this, any hashing strategy can be quite slow because of this. The fingerprint works very well for HPC environments.

rhpvorderman commented:
@cjllanwarne I have updated the documentation and the only test on Travis that fails is related to relative imports on CWL, so not something in this PR. I have tested the strategies in real life and found no problems. I see no "ready for review" label, and the "on-deck for review label" prioritizes the PR (which I feel I am not in a position to do). What is the usual process for declaring the PR ready?

illusional commented:
Hey @rhpvorderman, my testing isn't as comprehensive as yours but just wanted to weigh in and mention that the `fingerprint` method seems to be working locally, and on our Slurm cluster. I intend to do a bit more testing over the next few weeks but really appreciative of your work here.

cjllanwarne commented on CHANGELOG.md:
```diff
@@ -1,5 +1,16 @@
+
+#### new xxh64 and fingerprint strategies for call caching
+
+To alleviate problems with existing call cache strategies, two new strategies have been added: `xxh64` and 
```
```suggestion
To alleviate problems with existing call cache strategies, two new strategies for files on a shared filesystem (ie not in cloud storage) have been added: `xxh64` and 
```

cjllanwarne commented on docs/Configuring.md:
```diff
@@ -506,27 +507,52 @@ When `invalidate-bad-cache-results=true` (default: `true`), Cromwell will invali
 
 Cromwell also accepts [Workflow Options](wf_options/Overview#call-caching-options) to override the cache read/write behavior.  
 
+#### Call cache strategy options
```
These probably belong under the `Local filesystem options` section since they won't apply to cloud-hosted files which have their own hashing

cjllanwarne commented on docs/Configuring.md:
```diff
@@ -506,27 +507,52 @@ When `invalidate-bad-cache-results=true` (default: `true`), Cromwell will invali
+* Fingerprinting. This strategy works with containers.
+    * `fingerprint` tries to create a fingerprint for each file by taking its last modified time (milliseconds since
+       epoch in hexadecimal) + size (bytes in hexadecimal) + the xxh64 sum of the first 10 MB of the file. It is much
+       more lightweight than the hash based options while still unique enough that collisions are infeasible. This 
```
```suggestion
       more lightweight than the hash based options while still unique enough that collisions are unlikely. This 
```

cjllanwarne replied:
Certainly collisions are feasible, but under normal conditions (and unless you're trying to force it to happen!) I'd say vanishingly unlikely.

rhpvorderman replied:
Agreed. This is why I took the first 10 MB, not the first 1 MB. Some bioinformatics files have *very* large headers. More than 10 million characters though? 
You can imagine that in BAM files the headers are the same, and even the first aligned reads are the same, so the first 10 MB are the same. But then the total file size should also be the same, and the jobs should have finished in exactly the same millisecond. It is indeed quite unlikely.
I changed the wording.

cjllanwarne commented on docs/Configuring.md:
```diff
@@ -506,27 +507,52 @@ When `invalidate-bad-cache-results=true` (default: `true`), Cromwell will invali
+    * `path+modtime` creates a md5 hash of the path and its modification time.
+* Fingerprinting. This strategy works with containers.
+    * `fingerprint` tries to create a fingerprint for each file by taking its last modified time (milliseconds since
+       epoch in hexadecimal) + size (bytes in hexadecimal) + the xxh64 sum of the first 10 MB of the file. It is much
```
Was '10'MB based on anything specific or is it just a reasonable sounding number?

eg is there a good answer to "why not 5mb" or "why not 20mb"?

rhpvorderman replied:
Well 5MB and 20 MB are the same order of magnitude. So at that point it is arbitrary. However, "why not 1 MB?" or "why not 100 MB?" can be answered.  I have added a footnote explaining the choice.

cjllanwarne commented on docs/Configuring.md:
```diff
@@ -506,27 +507,52 @@ When `invalidate-bad-cache-results=true` (default: `true`), Cromwell will invali
+    * `fingerprint` tries to create a fingerprint for each file by taking its last modified time (milliseconds since
+       epoch in hexadecimal) + size (bytes in hexadecimal) + the xxh64 sum of the first 10 MB of the file. It is much
+       more lightweight than the hash based options while still unique enough that collisions are infeasible. This 
+       strategy works well for workflows that generate multi-gigabyte files and where hashing these files on the 
```
Given that cache copying might move files around, the timestamps will potentially be different for files which you want to treat as the same. So you might want to mention that this works best with specific cache copying strategies?

rhpvorderman replied:
Good point...
This works specifically well for hard-links, which do preserve the mod time.
However for copies you are correct. Here a new hash will have to be calculated. This is a good caveat to mention.

cjllanwarne commented on supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala:
```diff
@@ -87,10 +91,61 @@ final case class HashPathModTimeStrategy(checkSiblingMd5: Boolean) extends Confi
+    val hasher = xxhashFactory.newStreamingHash64(0)
+    val buffer: Array[Byte] = new Array[Byte](bufferSize)
+    var byteCounter: Long = 0
+    try {
```
No `catch` here makes me nervous...

Incidentally/additionally, you might find the `cromwell.util.TryWithResource` method useful.

rhpvorderman replied:
Added a catch statement here that catches all throwables and exceptions. Weirdly, Throwable does not inherit from Exception :confused: . So I just catch both now to be sure.

rhpvorderman replied:
Indeed I think the default behavior for `try ... finally` is to raise errors, but execute finally nevertheless. But I was not so sure about this crashing thing. It should raise errors when occurring obviously. I have removed `catch`.

cjllanwarne commented on supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala:
```diff
@@ -87,10 +91,61 @@ final case class HashPathModTimeStrategy(checkSiblingMd5: Boolean) extends Confi
+  override val description = "hash file content with md5"
 }
+
+final case class HashFileXxH64Strategy(checkSiblingMd5: Boolean) extends ConfigHashingStrategy {
```
I wonder whether it's possible to override the sibling md5 check for these strategies?

rhpvorderman replied:
Yes. All the sibling check does is read a string from a file with extension `.md5` it does not require any hashing at all. It is implemented in the ConfigHashingStrategy abstract class, so it also works for the new strategies.

cjllanwarne commented on supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala:
```diff
@@ -87,10 +91,61 @@ final case class HashPathModTimeStrategy(checkSiblingMd5: Boolean) extends Confi
+      file.lastModifiedTime.toEpochMilli.toHexString +
+      file.size.toHexString +
+      // Only check first 10 MB (10485760 bytes) for performance reasons
+      HashFileXxH64StrategyMethods.xxh64sum(file.newInputStream, maxSize = 10485760L)
```
Is it worth having a secondary digest to produce a fixed-length string. Right now this result is variable length which seems potentially a little strange

rhpvorderman replied:
:+1: Yes. I created a xxh64sumString method to calculate the xxh64 hash from a string. Should be implemented now.

cjllanwarne commented:
If you don't mind, I'd also suggest adding "This is a community feature" comments to the documentation to indicate that they're not supported by the core Cromwell team. Thanks!

rhpvorderman commented:
@cjllanwarne I have addressed your comments. I hope the changes are adequate. I have also added this feature as community supported in the docs, with no official support from the Cromwell core team.

rhpvorderman commented:
Tested the newer fingerprint hashing strategy (using a hash of equal size). Happy to report about 1000 cache hits correctly on our cluster.

illusional commented:
I've started testing this using the filebased db and it seems to be working as well. We're starting to scale to more users with `fingerprint` and seems to be working okay so far.

mcovarr commented on cromwell.example.backends/LocalExample.conf:
```diff
@@ -92,18 +92,25 @@
-              # Default: file
-              hashing-strategy: "file"
+              # Default: "md5"
+              hashing-strategy: "fingerprint"
```
I'm not sure how literally users "adapt" these example configs so I'm not sure about changing the choice here.

rhpvorderman replied:
Changed it back to md5.

rhpvorderman commented on supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala:
```diff
@@ -87,10 +91,77 @@ final case class HashPathModTimeStrategy(checkSiblingMd5: Boolean) extends Confi
+      HashFileXxH64StrategyMethods.xxh64sumString(file.lastModifiedTime.toEpochMilli.toHexString +
+      file.size.toHexString) +
+      // Only check first 10 MB (10485760 bytes) for performance reasons. 100 MB will take to much time on network file
+      // systems. 1 MB might not be unique enough.
```
@mcovarr I added the catch in response to https://github.com/broadinstitute/cromwell/pull/5450#discussion_r403260197 . But indeed I think the default behavior for `try ... finally` is ro raise errors, but execute finally nevertheless. But I was not so sure about this crashing thing. It should raise errors when occurring obviously. I have removed `catch`

rhpvorderman replied:
I made the size of the fingerprint configurable and added tests to ensure this works properly.

rhpvorderman commented on core/src/main/resources/reference_local_provider_config.inc.conf:
```diff
@@ -47,16 +47,25 @@ filesystems {
+      # For extended explanation check: https://cromwell.readthedocs.io/en/stable/Configuring/#call-caching
+      # "md5" will compute an md5 hash of the file content.
+      # "xxh64" will compute an xxh64 hash of the file content. Much faster than md5
+      # "fingerprint" will take last modified time, size and hash the first 10 mb with xxh64 to create a file fingerprint.
```
Yes, this was also duplicated in the example backend. So I had to edit it in 3 different places.
The reference conf and example backend conf now both suggest reading the documentation, so it only has to be changed in the documentation.

mcovarr commented on project/Dependencies.scala:
```diff
@@ -109,6 +110,7 @@ object Dependencies {
   private val workbenchModelV = "0.10-6800f3a"
   private val workbenchUtilV = "0.3-f3ce961"
 
+
```
🔙

mcovarr commented on supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala:
```diff
@@ -9,22 +9,31 @@ import cromwell.backend.standard.callcaching.StandardFileHashingActor.SingleFile
   def apply(hashingConfig: Config): ConfigHashingStrategy = {
       val checkSiblingMd5 = hashingConfig.as[Option[Boolean]]("check-sibling-md5").getOrElse(false)
 
+      // Fingerprint strategy by default checks the first 10 MB (10485760 bytes) for performance reasons.
```
nitpick: this value is 10 MiB (powers of 2) not MB (powers of 10)

mcovarr commented on supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala:
```diff
@@ -9,22 +9,31 @@ import cromwell.backend.standard.callcaching.StandardFileHashingActor.SingleFile
+      // Fingerprint strategy by default checks the first 10 MB (10485760 bytes) for performance reasons.
+      // 100 MB will take to much time on network file systems. 1 MB might not be unique enough.
+      // The value is user configurable.
+      lazy val fingerprintSize = hashingConfig.as[Option[Long]]("fingerprint-size").getOrElse(10485760L)
```
```suggestion
      lazy val fingerprintSize = hashingConfig.as[Option[Long]]("fingerprint-size").getOrElse(10 * (1 << 20))
```

rhpvorderman replied:
I will do 10 * 1024 * 1024 here as it is more clear than either 10485670 or `10 * (1 <<20)`

mcovarr commented on supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala:
```diff
@@ -87,10 +96,73 @@ final case class HashPathModTimeStrategy(checkSiblingMd5: Boolean) extends Confi
+      HashFileXxH64StrategyMethods.xxh64sum(file.newInputStream, maxSize = fingerprintSize)
+      }
+    }
+  override val description = "fingerprint the file with last modified time, size and a xxh64 hash of the first 10 mb"
```
description should be generalized for now-variable fingerprint size

rhpvorderman replied:
Now more general "first part of the file"

rhpvorderman commented:
@mcovarr Thanks for the review!

aednichols commented on CHANGELOG.md:
```diff
@@ -1,5 +1,16 @@
+
+#### new xxh64 and fingerprint strategies for call caching
+
+To alleviate problems with existing call cache strategies, two new strategies for files on a shared filesystem (ie not in cloud storage) have been added: `xxh64` and 
```
Can you be specific on what problems are being alleviated and how? Would the following be correct?
```suggestion
Existing call cache strategies can be slow on shared filesystems (SFS backend, i.e. not in cloud storage). Two faster strategies have been added for this use case: `xxh64` and 
```

rhpvorderman replied:
Well `path` and `path+modtime` are already extremely fast, but they don't work with docker. Md5 is the slow strategy. Thanks for the suggestion. I have updated the description.

aednichols commented:
@rhpvorderman this is good to merge as soon as CI passes (I took the liberty of resolving a trivial changelog conflict on your behalf)

cjllanwarne commented:
Thank you, @rhpvorderman !

rhpvorderman commented:
Thanks everyone for the feedback and the merging!
